### PR TITLE
Fix Muon optimizer conflict with gradient clipping in ZeRO 1/2

### PR DESCRIPTION
**Pull request by fy817**
### Summary

This PR fixes a critical logic error in how gradient clipping interacts with the **Muon optimizer** within the ZeRO Stage 1 and 2 implementations.

Currently, the `muon_update` (orthogonalization) is applied before gradient clipping. Since Muon's orthogonalization naturally changes the gradient magnitude (often resulting in norms significantly larger than the default `clip_grad` of 1.0), the subsequent standard clipping mechanism incorrectly identifies these valid updates as exploding gradients. This results in aggressive, erroneous clipping that severely degrades or breaks training.

This fix ensures that gradient clipping is applied to the raw gradients **before** the Muon update occurs, ensuring the orthogonalization step receives properly clipped inputs, while preventing double-clipping during the optimizer step.

### Root Cause

In the current implementation, `muon_update` is invoked inside `get_flat_partition` to leverage the shaped gradients required by the algorithm. However, `get_flat_partition` is called before the standard global gradient clipping phase.

1. **Muon Characteristics:** The Muon optimizer performs an orthogonalization step (via Newton-Schulz iteration), which typically results in gradient tensors with Frobenius norms much larger than 1.0.
2. **Clipping Conflict:** DeepSpeed defaults to a `clip_grad` value of 1.0. Because the gradients have already been processed by Muon by the time clipping occurs, their norms are naturally high.
3. **Result:** The global clipper perceives these high norms as gradient explosion and scales the gradients down by a massive factor (e.g., dividing by 100+), effectively neutralizing the Muon update and causing training failure.

### Fix

The fix rearranges the clipping logic for parameters utilizing the Muon optimizer:

1. **Pre-computation:** In `independent_gradient_partition_epilogue`, we now pre-calculate the unscaled global gradient norm (or the specific norm required for clipping) before the flattening process begins.
2. **Pre-Update Clipping:** This calculated norm is passed into `get_flat_partition`. Inside this method, we calculate the `clip_factor` and apply it (`grad = grad / clip_factor`) explicitly to the gradients **before** passing them to `muon_update`. This ensures Muon operates on clipped, stable gradients.
3. **Skip Redundant Clipping:** In the `step` method, we modified the call to `unscale_and_clip_grads` to accept a `skip_clipping=True` flag. For Muon parameter groups, we set this flag to true, ensuring that we only perform unscaling (handling loss scale) without re-applying gradient clipping, as it was already handled in the previous step. Standard AdamW parameters retain the original behavior.

### CODE REVIEW

**Comment by sfc-gh-truwase** [comment]
@PKUWZP, please help review

**Comment by fy817** [comment]
Hi @PKUWZP, could you please take a look when you have a moment? Let me know if you need anything else from my side. Thanks!

**Comment by fy817** [comment]
Hi @sfc-gh-truwase, it seems @PKUWZP might be busy at the moment. Would it be possible for you or another maintainer to take a look? Thanks for your help!

**Comment by sfc-gh-truwase** [comment]
@fy817 can you please resolve the conflict?

**Comment by fy817** [comment]
@sfc-gh-truwase Thank you for the feedback. The conflicts have been resolved. Please refer to my comment above for details on how I handled the conflicts. Looking forward to your review.

**Comment by sfc-gh-truwase** [comment]
> The conflicts have been resolved. Please refer to my comment above for details on how I handled the conflicts. Looking forward to your review

Can you please clarify how the conflicts were resolved, since GH is still reporting a conflict
image

**Comment by fy817** [comment]
@sfc-gh-truwase  Sorry about that! I misunderstood the context and missed the merge conflict. It's been fixed now. Thanks for the heads-up.

**Comment by fy817** [comment]
@sfc-gh-truwase Hi, I noticed that this PR has been linked with #7808. Please let me know if there is anything else needed from my side. I am happy to help get this merged.

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 1]
```diff
@@ -859,6 +859,32 @@ def independent_gradient_partition_epilogue(self):
             self._clear_previous_reduced_grads()
 
         if self.cpu_offload is False:
+            # Pre-compute gradient norm for Muon clipping if needed
+            grad_norm_for_muon = None
+            if self.is_gradient_accumulation_boundary:
+                # Check if any parameter group uses Muon
+                uses_muon = False
```
Comment (sfc-gh-truwase) [inline deepspeed/runtime/zero/stage_1_and_2.py:866]: The main connection to #7808 is that `use_muon` should be an object field set in the constructor and defined per-param_group.

Comment (PKUWZP) [inline deepspeed/runtime/zero/stage_1_and_2.py:866]: @fy817 Like @sfc-gh-truwase mentioned, `use_muon` should be an object field. This should be computed once and stored as a class field.  Computing this repeatedly is inefficiency and can cause errors.

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 2]
```diff
@@ -859,6 +859,32 @@ def independent_gradient_partition_epilogue(self):
...
+                for i, _ in enumerate(self.bit16_groups):
+                    if len(self.params_in_partition[i]) > 0 and getattr(self.params_in_partition[i][0], 'use_muon', False):
```
Comment (PKUWZP) [inline deepspeed/runtime/zero/stage_1_and_2.py:868]: @fy817 `getattr(tensor_list[0], 'use_muon', False)` needs to be set externally by the user. Otherwise there's no documentation and validation that this attribute would exist or be set properly.

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 3]
```diff
@@ -859,6 +859,32 @@ def independent_gradient_partition_epilogue(self):
...
+                        uses_muon = True
```
Comment (PKUWZP) [inline deepspeed/runtime/zero/stage_1_and_2.py:869]: @fy817 One suggestion is that add validation in `__init__ ` of the `DeepSpeedZeroOptimizer` class that checks for use_muon attributes when a Muon optimizer is detected, and add documentation about this requirement.

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 4]
```diff
@@ -859,6 +859,32 @@ def independent_gradient_partition_epilogue(self):
             self._clear_previous_reduced_grads()
 
         if self.cpu_offload is False:
+            # Pre-compute gradient norm for Muon clipping if needed
+            grad_norm_for_muon = None
```
Comment (PKUWZP) [inline deepspeed/runtime/zero/stage_1_and_2.py:863]: Here you introduced a new instance variable `grad_norm_for_muon` that's set in `independent_gradient_partition_epilogue` function and consumed in `get_flat_partition` function. This would potentially cause data racing problem due to the resulting implicit dependency on the call orders. One solution is that we make the `grad_norm_for_muon` defined explicitly outside the function, then we can ensure that it's always valid before being passed into functions.

**Comment by PKUWZP** [comment]
@fy817 Thanks so much for submitting the PR. I have reviewed the PR and left a few comments. One fundamental question I have is: do we actually need "naive gradient clipping" for Muon Optimizer? I feel that it might not even needed for Muon. Here are my thoughts: 

In Muon Optimizer Implementation (shown below), the Newton-Schulz iteration already normalizes by spectral norm internally. The output is approximately an orthogonal matrix whose norm is determined by the matrix shape, not the input gradient magnitude.

  `def zeropower_via_newtonschulz5(G, steps: int):`
          `X = G.bfloat16()`
          `# Ensure spectral norm is at most 1`
         `X = X / (X.norm(dim=(-2, -1), keepdim=True) + 1e-7) `
        ` # Newton-Schultz Iterations`
         `for _ in range(steps):`
             `A = X @ X.mT`
             `B = b * A + c * A @ A`
             `X = a * X + B @ X`
         `return X`

  This means:
  1. Exploding gradients are already handled - the spectral norm division bounds the input
  2. The output norm is shape-dependent, not gradient-dependent - no need to clipping the output
  3. Performing pre-clipping changes gradient direction unnecessarily - and the magnitude will be normalized anyway

Instead, I think MuonClilp (described in Kimi K2 technical report: https://arxiv.org/abs/2507.20534 ) is actually interesting for this case. It rescales query and key matrices within attention layers after each update to further improve training stability. I just don't feel doing naive pre-clipping is needed here for Muon. 

Also in this PR, the global norm is computed via `scaled_global_norm()` on all `averaged_gradients`, this is not suitable for mixed setup (e.g. Muon for hidden layers while AdamW for embedding layers ). If you do this, the global norm will include the orthogonalized Muon gradients, which would not be the right thing to do to clip the AdamW parameter groups. Thoughts?

**Review by PKUWZP — COMMENTED** [review]
@fy817 Thanks so much for submitting the PR. I have reviewed the PR and left a few comments. One fundamental question I have is: do we actually need "naive gradient clipping" for Muon Optimizer? I feel that it might not even needed for Muon. Here are my thoughts: 

In Muon Optimizer Implementation (shown below), the Newton-Schulz iteration already normalizes by spectral norm internally. The output is approximately an orthogonal matrix whose norm is determined by the matrix shape, not the input gradient magnitude.

  `def zeropower_via_newtonschulz5(G, steps: int):`
          `X = G.bfloat16()`
          `# Ensure spectral norm is at most 1`
         `X = X / (X.norm(dim=(-2, -1), keepdim=True) + 1e-7) `
        ` # Newton-Schultz Iterations`
         `for _ in range(steps):`
             `A = X @ X.mT`
             `B = b * A + c * A @ A`
             `X = a * X + B @ X`
         `return X`

  This means:
  1. Exploding gradients are already handled - the spectral norm division bounds the input
  2. The output norm is shape-dependent, not gradient-dependent - no need to clipping the output
  3. Performing pre-clipping changes gradient direction unnecessarily - and the magnitude will be normalized anyway

Instead, I think MuonClilp (described in Kimi K2 technical report: https://arxiv.org/abs/2507.20534 ) is actually interesting for this case. It rescales query and key matrices within attention layers after each update to further improve training stability. I just don't feel doing naive pre-clipping is needed here for Muon. 

Also in this PR, the global norm is computed via `scaled_global_norm()` on all `averaged_gradients`, this is not suitable for mixed setup (e.g. Muon for hidden layers while AdamW for embedding layers ). If you do this, the global norm will include the orthogonalized Muon gradients, which would not be the right thing to do to clip the AdamW parameter groups. Thoughts?

**Comment by fy817** [comment]
@sfc-gh-truwase, @PKUWZP, thank you for your responses. Below are some of my thoughts:
1. Indeed, the Muon optimizer may not require gradient clipping. My experiments also confirm this—setting the gradient clipping value to 0 resolves the aforementioned conflict between the Muon optimizer and gradient clipping, and normal training performance can still be achieved.
2. However, I would like to point out that the conflict between the Muon optimizer and gradient clipping still exists in the current implementation. First, users might overlook the gradient clipping setting. Since DeepSpeed defaults to a gradient clipping value of 1.0, this can lead to incorrect gradient modification due to clipping, ultimately causing training failures—a problem that still needs to be addressed. Second, even if we resolve the conflict by setting gradient clipping to 0, while Muon itself may not need clipping, users might still require gradient clipping for the Adam update portion of the optimizer.
3. Thank you very much for pointing out the shortcomings in the current PR. I look forward to further discussion to refine this solution.
